packages/@aws-cdk/core-synthesizer/lib/staging-stack.ts

```diff
@@ -0,0 +1,184 @@
+import * as ecr from '@aws-cdk/aws-ecr';
+import * as iam from '@aws-cdk/aws-iam';
+import * as s3 from '@aws-cdk/aws-s3';
+import { Arn, ArnFormat, Aws, DockerImageAssetSource, FileAssetSource, RemovalPolicy, Stack, StackProps } from '@aws-cdk/core';
+import { Construct, IConstruct } from 'constructs';
+
+export interface FileAssetInfo {
+  readonly bucketName: string;
+  readonly assumeRoleArn: string;
+}
+
+export interface DockerAssetInfo {
+  readonly repoName: string;
+  readonly assumeRoleArn: string;
+}
+/**
+ * Information on how a Staging Stack should look.
+ */
+export interface IStagingStack extends IConstruct {
+  /**
+   * The app-scoped, environment-keyed bucket created in this staging stack.
+   */
+  readonly stagingBucket?: s3.Bucket;
+
+  /**
+   * The app-scoped, environment-keyed repositories created in this staging stack.
+   * A repository is created per image asset family.
+   */
+  readonly stagingRepos: Record<string, ecr.Repository>;
+
+  /**
+   * // TODO
+   */
+  addFile(asset: FileAssetSource): FileAssetInfo;
+
+  /**
+   * // TODO
+   */
+  addDockerImage(asset: DockerImageAssetSource): DockerAssetInfo;
+}
+
+/**
+ * Staging Stack Properties
+ */
+export interface StagingStackProps extends StackProps {
+  /**
+   * Explicit name for the staging bucket
+   *
+   * @default - a well-known name unique to this app/env.
+   */
+  readonly stagingBucketName?: string;
+
+  /**
+   * Pass in an existing role to be used as the file publishing role.
+   *
+   * @default - a well-known name unique to this app/env.
+   */
+  readonly fileAssetPublishingRoleName?: string;
+
+  /**
+   * Pass in an existing role to be used as the image publishing role.
+   *
+   * @default - a well-known name unique to this app/env.
+   */
+  readonly dockerAssetPublishingRoleName?: string;
+}
+
+/**
+ * A default Staging Stack
+ */
+export class DefaultStagingStack extends Stack implements IStagingStack {
+  /**
+   * Default asset publishing role name for file (S3) assets.
+   */
+  private static readonly DEFAULT_FILE_ASSET_PUBLISHING_ROLE_NAME = 'cdk-${Qualifier}-file-publishing-role-${AWS::AccountId}-${AWS::Region}';
```

Comment:
This name is going to conflict with the default bootstrapped roles :). 

I think we pick a prefix that contains the app name and base the rest off of it. And come to think of it, `${AccountId}` only needs to be in the bucket name, not in any of the role names.